app/models/broadcast.rb

```diff
@@ -5,10 +5,28 @@ class Broadcast < ApplicationRecord
 
   validates :title, :type_of, :processed_html, presence: true
   validates :type_of, inclusion: { in: %w[Announcement Welcome] }
+  validate  :single_active_announcement_broadcast
 
   scope :active, -> { where(active: true) }
+  scope :announcement, -> { where(type_of: "Announcement") }
+  scope :welcome, -> { where(type_of: "Welcome") }
 
   def get_inner_body(content)
     Nokogiri::HTML(content).at("body").inner_html
   end
+
+  private
+
+  def single_active_announcement_broadcast
+    # Only add errors if we are trying to modify an announcement
+    # broadcast while another announcement broadcast is already
+    # active to ensure that only one can be active at a time.
+    active_broadcasts = Broadcast.announcement.active
+    return unless active_broadcasts.count.positive? &&
+      active &&
+      type_of == "Announcement" &&
+      active_broadcasts.first.id != id
```

Comment:
super small optimization since they are in `AND`: if you put the count as the third condition, it's going to be executed only if the previous ones are all true. This way we don't hit the DB with a SQL count if we don't need to

You might also not need the count at all actually, though it's going to make the code less readable, since the last condition loads the first item just to check the `ID`, so instead of going to the DB twice, you can do (untested): 

```ruby
active && type_of == "Announcement" && ![nil, id].include?(active_broadcasts.order(id: :asc).limit(1).pluck(:id).first)
```

a cleaner version could be:

```ruby
first_active_broadcast = active_broadcasts.order(id: :asc).limit(1) # this doesn't execute a query
return unless active && type_of == "Announcement" &&  ![nil, id].include?(first_active_broadcast.pluck(:id).first)
```


which in Rails 6 will become:

```ruby
first_active_broadcast = active_broadcasts.order(id: :asc).limit(1) # this doesn't execute a query
return unless active && type_of == "Announcement" &&  ![nil, id].include?(first_active_broadcast.pick(:id))
```

:-D

Reply:
@rhymes Ooof yes I should have known to put the query last to avoid unnecessary optimizations. This is what happens when I try to code in the evenings 🙈